src/Microsoft.ApplicationInsights/Extensibility/Implementation/Endpoints/EndpointProvider.cs

```diff
@@ -0,0 +1,180 @@
+                    catch (UriFormatException ex)
+                    {
+                        throw new ConnectionStringInvalidEndpointException($"The connection string endpoint is invalid. EndpointName: {endpointName} EndpointProperty: {endpointMeta.ExplicitName}", ex);
+                    }
+                }
+                else if (this.connectionStringParsed.TryGetValue("EndpointSuffix", out string endpointSuffix))
+                {
+                    try
+                    {
+                        return BuildUri(
+                            prefix: endpointMeta.EndpointPrefix,
+                            suffix: endpointSuffix,
+                            location: this.GetLocation());
+                    }
+                    catch (UriFormatException ex)
+                    {
+                        throw new ConnectionStringInvalidEndpointException($"The connection string endpoint is invalid. EndpointName: {endpointName} Either EndpointSuffix or Location.", ex);
+                    }
+                }
+                else
+                {
+                    return new Uri(endpointMeta.Default);
+                }
+            }
+            catch (Exception ex)
+            {
+                // TODO: LOG TO ETW ERROR: exception trying to get endpointName. Log Inner Exception.
+                throw; // Re-throw original exception
+            }
+        }
+
+        /// <summary>
+        /// Will evaluate connection string and return the requested instrumentation key.
+        /// </summary>
+        /// <returns>Returns the instrumentation key from the connection string.</returns>
+        public string GetInstrumentationKey()
+        {
+            if (this.connectionStringParsed.TryGetValue("InstrumentationKey", out string value))
+            {
+                return value;
+            }
+            else
+            {
+                throw new ConnectionStringMissingInstrumentationKeyException();
+            }
+        }
+
+        /// <summary>
+        /// Parse a string connection string and return a Dictionary.
+        /// </summary>
+        /// <remarks>Example: "key1=value1;key2=value2;key3=value3".</remarks>
+        /// <returns>A dictionary parsed from the input connection string.</returns>
+        internal static Dictionary<string, string> ParseConnectionString(string value)
+        {
+            if (string.IsNullOrEmpty(value))
+            {
+                return new Dictionary<string, string>(0);
+            }
+
+            try
+            {
+                return value
+                    .Split(SplitSemicolon, StringSplitOptions.RemoveEmptyEntries)
+                    .Select(part => part.Split('='))
+                    .ToDictionary(split => split[0], split => split[1], StringComparer.OrdinalIgnoreCase);
+            }
+            catch (ArgumentException ex) when (ex.Message.StartsWith("An item with the same key has already been added.", StringComparison.Ordinal))
```

Comment:
does custom exception type worth this logic?

If it does, I suggest to avoid creating particular exception based on the dictionary exception message and fill  dictionary without LINQ and with Contains checks.

Reply:
I really shouldn't have to say this: Exception message can be localized. Don't filter exceptions based on their message!

Reply:
Crap, I didn't consider localization. :( 
I'll replace this tomorrow.

Reply:
Even if there is no localization, determine the exception type by parsing exception message is a very dangerous idea - the fact that you're relying on the actual implementation behavior rather than the interface.